new literacy::{Read, Write} traits to handle std/no_std (alternative)

RCasatta opened the pull request:
Alternative to https://github.com/rust-bitcoin/bitcoin_hashes/pull/126
see description there

TheBlueMatt commented on src/literacy.rs:
```diff
@@ -0,0 +1,123 @@
+#[cfg(all(feature = "std", feature = "use-core2"))]
+compile_error!("feature \"std\" and \"use-core2\" cannot be enabled together.");
```
I think we should "default" to core2 in this case - if we already are building against the core2 dep, we might as well just assume the user set core2/std and then the core2 impl is the same as doing the std impl.

RCasatta replied:
I agree but there are some implementation issue, in particular

> we might as well just assume the user set core2/std

How so? with another feature `use-core2-std = ["core2/std"]`? This would cause other issues in feature combinations...
The other option, using `std = ["core2/std"]` will cause problem in default features compilation on rust 1.29

TheBlueMatt replied:
> How so? with another feature use-core2-std = ["core2/std"]? This would cause other issues in feature combinations...

We don't have to specify it ourselves for core2/std to be enabled by some other dependency path. Its probably neater to just impl for std and not core2 than to compile-error, and if users intended to use core2 they can always fix it by, themselves, directly requiring core2/std. This avoids the issue of having some crate that depends on A and B and A enables std and B enables core2, giving the user no way to fix the issue aside from patching either A or B.

RCasatta replied:
Sure, removed the `compile_error` and implementing std impl only if there is std but not core2

TheBlueMatt commented on src/literacy.rs:
```diff
@@ -0,0 +1,123 @@
+mod std_impl {
+    use super::{Read, Write};
+
+    impl<R: ::std::io::Read> Read for R {
```
Is this what we want? Don't we want to `impl<R: OurReadTrait> std::io::Read for R` instead? That way a user can say `Hasher as std::io::Read`?

TheBlueMatt replied:
Right, I'd managed to confuse myself between Read and Readable-style traits. In general I agree this is correct, but I do feel like its somewhat opposite of what we want for bitcoin_hashes. Maybe this is correct for Read but we should swap it for Write?

TheBlueMatt replied:
> I feel the same for hashes, but thinking about rust-bitcoin and the encoding traits that would be annoying again.

Yea, I guess I'm ok with it as-is.

> Thinking about it, what keeps us from just implementing both std::io::Write and core2::Write for hash engines directly (depending on features), just calling the update fn.

For bitcoin_hashes, I don't disagree, but we'll want this stuff for rust-bitcoin (and rust-lightning), so I think its best to do it here and then use the same traits downstream.

TheBlueMatt commented:
I think also we should have a not-std&&not-core2 impl block for Write into a Vec and Read from a slice, like @devrandom did in rust-bitcoin at https://github.com/rust-bitcoin/rust-bitcoin/pull/603/files#diff-76866598ce8fd16261a27ac58a84b2825e6e77fc37c163a6afa60f0f4477e569R199-R256

RCasatta commented on contrib/test.sh:
```diff
@@ -55,20 +48,19 @@ if [ "$DO_WASM" = true ]; then
     cargo clean
     CC='clang -fsanitize=memory -fno-omit-frame-pointer'                                         \
     RUSTFLAGS='-Zsanitizer=memory -Zsanitizer-memory-track-origins -Cforce-frame-pointers=yes'   \
-    cargo test --lib --all --features="$FEATURES" -Zbuild-std --target x86_64-unknown-linux-gnu
```
should be `--all-features` here and above, waiting to find a solution for defaulting to `core2` if enabled (without conflicting with `std`)

RCasatta commented:
> I think also we should have a not-std&&not-core2 impl block for Write into a Vec and Read from a slice, like @devrandom did in rust-bitcoin at https://github.com/rust-bitcoin/rust-bitcoin/pull/603/files#diff-76866598ce8fd16261a27ac58a84b2825e6e77fc37c163a6afa60f0f4477e569R199-R256

added in 4291b4b and also removed write_all blanket impl

TheBlueMatt commented on src/lib.rs:
```diff
@@ -25,7 +25,7 @@
 #![deny(non_snake_case)]
 #![deny(unused_mut)]
-#![deny(missing_docs)]
+//#![deny(missing_docs)]
```
I hope you didn't intend to land this :)

RCasatta replied:
removed, and added doc where missing in 8bf2f9ed2fd5f6be0693b502e9a5351edef300d4

TheBlueMatt reviewed (commented):
I think this is basically almost there. Is it time to un-draft it?

sgeisler reviewed (commented):
utACK 8bf2f9ed2fd5f6be0693b502e9a5351edef300d4 for the rust code, not so sure about the tests, needs some clarification at least imo.

TheBlueMatt commented:
> `impl<W: ::std::io::Write> Write for &mut W {`

What about `impl<W: ::literacy::Write> ::literact::Write for &mut W {`? That's closer to what `std::io::Write` does, though I have a feeling that will conflict, too, because of the blanket `std::io::Write` impl. I don't think its the end of the world to move rust-bitcoin to always taking a `&mut W` instead, though. @apoelstra ?

devrandom commented:
Yeah, the other way results in the same conflict.

I'm playing with taking `&mut W` instead of `mut W`, and that seems to work.

A further issue is that rust-bitcoin uses `Read::take`, which we don't implement here.  Should we grab that from std?

And one more question - how exactly do we concretize the `Error` type param?  For example, is this what we want for the `Encodeable` trait in rust-bitcoin (and the few dozen impls)?

```rust
pub trait Encodable {
    fn consensus_encode<W: io::Write<Error=io::Error>>(&self, writer: &mut W) -> Result<usize, io::Error>;
}
```

(where the `io::Error` type depends on std vs no-std)

sgeisler commented:
I don't really like the `BorrowMut<dyn (::std::io::Write)>` solution, do we know if it will be optimized away? Or will people unknowingly be using dynamic dispatch? Especially for something kinda low-level like io that would be annoying.

I think implementing `literacy::Write` and `literacy::Read` directly is actually an outlier (although very prominent in hashes). Afaik the remaining API of `rust-bitcoin` is more a consumer of objects implementing these traits. Since `std` implements `impl<W: Write + ?Sized> Write for &mut W` and we implement `impl<W: ::std::io::Write> Write for W` we already get `literacy` impls for most relevant mut borrows, making the hacky blanket `BorrowMut` impl unnecessary except for the few cases where `literacy` traits are implemented directly. I propose to instead use a macro to just impl `literacy::Write` for both `HashEngine` and `&mut HashEngine` (19b74b75edfa7e5d2107b3be86c17d9ed36049ac).

afilini commented:
> making the hacky blanket BorrowMut impl unnecessary except for the few cases where literacy traits are implemented directly.

Not even that, because implementing it that way made it only apply to types that already implemented `std::io::Write` and that we were "bridging" to `literacy::Write`, which as you pointed out already had an implementation for their references since the std trait is implemented on both automatically. We were just trying to come up with something cool but in the end I don't think it can be done automatically.

TheBlueMatt commented:
Yea, I think we definitely can't do anything that relies on dyn here, the optimization story of this type of deserialization logic is already pretty rough (and we've had users with significant performance bottlenecks in our deserialization before, it probably still is the bottleneck for electrs), we can't make it worse.

RCasatta commented:
I think 19b74b7 is the way to go, but there are some issues with lifetimes on 1.29 https://github.com/RCasatta/bitcoin_hashes/runs/2631403404?check_suite_focus=true, @sgeisler do you want to look at it?

sgeisler commented:
Sure, will do, the commit was merely to illustrate my point, I'll also remove the example again.

sgeisler commented:
The embedded test broke somehow :frowning: I'm not sure if it has anything to do with my commit. Did it increase the lib size so much that it doesn't fit into the flash anymore?

RCasatta commented:
> The embedded test broke somehow 😦 I'm not sure if it has anything to do with my commit. Did it increase the lib size so much that it doesn't fit into the flash anymore?

Checked on my machine and the compiled size didn't change... Turns out it is related to the latest nightly compiler (embedded test runs with nightly) because with https://github.com/RCasatta/bitcoin_hashes/commit/095d42304b2b9d170735f10286f81eb5f3d9f2fd it works https://github.com/RCasatta/bitcoin_hashes/actions/runs/861640382

Tomorrow I will try to fix the comments about tests from @sgeisler ...

RCasatta commented on contrib/test.sh:
```diff
@@ -18,35 +19,27 @@ export CARGO_TERM_VERBOSE=true
+          echo "--------------$feature----------------"
+          cargo build --no-default-features --features="$feature"
+          if [[ ${feature} =~ "std," ]] ; then
+              cargo test --no-default-features --features="$feature"
```
because tests have std enabled

`#![cfg_attr(all(not(test), not(feature = "std")), no_std)]`

RCasatta commented on contrib/test.sh:
```diff
@@ -1,10 +1,11 @@
-FEATURES="serde serde-std"
+# Combination of features to test, should be every features combination but std and core2 together
+# note std has a comma in the end so that following regex avoid matching serde-std
+FEATURES=("" "std," "use-core2" "use-core2-std" "std,use-core2" "std,serde-std" "use-core2,serde-std")
```
serde added in 7b909c986f4490411c2773900108a91e115f922c, I forget about it because it's used but not listed in the Cargo.toml feature section...

Not sure if other combination are needed though

RCasatta commented:
> A further issue is that rust-bitcoin uses Read::take, which we don't implement here. Should we grab that from std?

added in a37f2bed96697b0f8fac04b217e6911bc1ba8417

Another option without adding the fn to the trait and the associated type would have been to restore the `CappedRead` https://github.com/RCasatta/rust-bitcoin/commit/24015e8ed17a4d880311247cba2b58b53deaa890, however I preferred like it's done a37f2bed96697b0f8fac04b217e6911bc1ba8417 (inheriting std/core2 implementation, and the simple implementation for the slice, there should not be any other case where we implement Read in rust-bitcoin if I am not mistaken)

devrandom commented:
This question remains, how do we concretize `Error` for `Encodable` in rust-bitcoin?

One issue is that the hash engines already concretize `Error` to `()` in this crate, so that would require a bunch of error mapping if `Encodable` wants to concretize to `io::Error`.

> And one more question - how exactly do we concretize the `Error` type param? For example, is this what we want for the `Encodable` trait in rust-bitcoin (and the few dozen impls)?
> 
> ```rust
> pub trait Encodable {
>     fn consensus_encode<W: io::Write<Error=io::Error>>(&self, writer: &mut W) -> Result<usize, io::Error>;
> }
> ```
> 
> (where the `io::Error` type depends on std vs no-std)

RCasatta commented:
> This question remains, how do we concretize Error for Encodable in rust-bitcoin?

with the latest commits it should be:
(literacy::Error it's a re-export according to the activated features)
```
  pub trait Encodable {
      fn consensus_encode<W: literacy::Write<Error=literacy::Error>(&self, writer: &mut W) -> Result<usize, literacy::Error>;
  }
```

> One issue is that the hash engines already concretize Error to () in this crate, so that would require a bunch of error mapping if Encodable wants to concretize to io::Error.

1f500f1e41c7d29f914558b3b1c73ab101b61610 use literacy::Error also for engines (the `()` was conveying more the idea of the absence of error but this way should be more convenient to use and I added a comment)

afilini commented on src/literacy.rs:
```diff
@@ -0,0 +1,263 @@
+    /// The error type returned in Result
+    type Error;
+    /// The type to implement limited reads
+    type Take;
```
I would add a trait bound to literacy::Read here, it might help later on when working with generics

afilini replied:
Alternatively you could also add another trait specifically for the Take type that "extends" `Read` and forces a specific constructor-like method. With that you can then provide a default implementation for `take()` like std does, because you can always construct an object of type `Self::Take` using the constructor defined by its trait.

RCasatta replied:
added trait bound in 6b7b28fc3536fda034d70d39a5d1bded78f07670

RCasatta commented:
After adding the `literacy::Error` logic I realized it may be possible to remove the associated parameter Error, adding something like alloc::Box<dyn OurErrorTrait> where OurErrorTrait: Display + Debug + … {} for the default_impl...

But I would like to ear opinions before doing such changes

TheBlueMatt commented:
> This question remains, how do we concretize Error for Encodable in rust-bitcoin?

How much do we actually add an error ourselves? I assume we should basically never error unless the underlying `Write` does, so could we not just return `Result<usize, W::Error>`?

devrandom commented:
Do we really want these verbose function signatures (illustrated for the std case):

```rust
pub trait Decodable: Sized {
    /// Decode an object with a well-defined format
    fn consensus_decode<D: io::Read<Error=Error, Take=::std::io::Take<D>>>(d: &mut D) -> Result<Self, Error>;
}

impl Decodable for OutPoint {
    fn consensus_decode<D: io::Read<Error=encode::Error, Take=::std::io::Take<D>>>(d: &mut D) -> Result<Self, encode::Error> {
        Ok(OutPoint {
            txid: Decodable::consensus_decode(d)?,
            vout: Decodable::consensus_decode(d)?,
        })
    }
}
```

sgeisler commented:
> After adding the `literacy::Error` logic I realized it may be possible to remove the associated parameter Error, adding something like `alloc::Box<dyn OurErrorTrait>` where `OurErrorTrait: Display + Debug + … {}` for the default_impl...

Please don't, the entire point of this PR was to have the error be an associated type instead of a `Box<dyn …>` compared to #126 .

> This question remains, how do we concretize Error for Encodable in rust-bitcoin?

My vision for the API is as follows:

```rust
pub trait Encodable {
    fn consensus_encode<W: literacy::Write>(&self, writer: W) -> Result<usize, W::Error>;
}

pub trait Decodable: Sized {
    fn consensus_decode<D: literacy::Read>(d: D) -> Result<Self, DecodeError<D::Error>>;
}

pub enum DecodeError<E: Display + Debug + …> {
    Io(E),
    Psbt(psbt::Error),
    …
}
```

The caller of one of these functions typically knows if they use std or core2, or if not they will just have to do the same generic trick that we are doing here. So there is no need to use a `Box<dyn>` for errors.

> Do we really want these verbose function signatures (illustrated for the std case):

I think there are some misconceptions about having to name associated types in that context, you don't have to. You can access them, for every struct-trait pair these are well-defined, so no need to actually specify them anywhere here.

RCasatta commented:
@TheBlueMatt @sgeisler so it's good like it is?

sgeisler commented on src/literacy.rs:
```diff
@@ -0,0 +1,263 @@
+    /// see [std::io::Read::read_exact]
+    fn read_exact(&mut self, buf: &mut [u8]) -> ::core::result::Result<(), Self::Error>;
+    /// see [std::io::Read::take]
+    fn take(self, limit: u64) -> Self::Take;
```
Isn't it more idiomatic to take a `usize`?

RCasatta replied:
I just did the same as  https://doc.rust-lang.org/std/io/trait.Read.html#method.take

sgeisler replied:
I guess not looking at std, weird.

devrandom commented on src/literacy.rs:
```diff
@@ -0,0 +1,263 @@
+    /// see [std::io::Read::read_exact]
+    fn read_exact(&mut self, buf: &mut [u8]) -> ::core::result::Result<(), Self::Error>;
+    /// see [std::io::Read::take]
+    fn take(self, limit: u64) -> Self::Take;
```
This doesn't seem to quite work with the `&mut` pattern:

```
    fn consensus_decode<D: io::Read>(d: &mut D) -> Result<Self, encode::Error> {
        let mut d = d.take(MAX_VEC_SIZE as u64);
```

gives:

```
error[E0507]: cannot move out of `*d` which is behind a mutable reference

    let mut d = d.take(MAX_VEC_SIZE as u64);
                      ^ move occurs because `*d` has type `D`, which does not implement the `Copy` trait
```

RCasatta replied:
maybe removing `take` from the trait and using something like the CappedRead in https://github.com/RCasatta/rust-bitcoin/commit/24015e8ed17a4d880311247cba2b58b53deaa890 ?

devrandom replied:
I'll give this a try.

devrandom replied:
Seems to work fine, let's take out the `take`.

RCasatta replied:
take removed in c6d4ca48fd56189a1c20b08c8fd81786b3b65d74

sgeisler commented:
I tried to migrate the consensus de/encoding API in rust-bitcoin to this PR and have to say it's not too pleasant :laughing: some problems I noticed:

* The generic error type is infecting way more parts of the code than I expected, partly because the `encode::Error` is not limited to io-related functionality. I think it would actually be a good thing to untangle it a bit, but that might turn out to be a very big task.
* Some of our code actually uses the information in the io errors, e.g. there is `StreamReader::read_next` which tries to decode some data to see if it received enough bytes or if it needs more (being signaled by `io::ErrorKind::UnexpectedEof`).

This means we'd have to do a lot of work to make `rust-bitcoin` compatible. But I also noticed that the `ErrorKind` type from both `std` and `core` is the same. So a middle ground might be found by building our own error type that is slightly more intelligent than `Box<dyn OurErrorTrait>`:

```rust
struct IoError {
    kind: IoErrorKind,
    error: Box<dyn OurErrorTrait>, //
}

trait OurErrorTrait: Debug + Any {}
```

This would allow cheap conversion from both `std` and `core2` errors, cheap superficial mathching of error causes and expensive deeper introspection when needed by downcasting the underlying error. It's not pretty, but probably a lot less work than restructuring the entire error handling of `rust-bitcoin`. What do you think?

RCasatta commented:
Would like to ear from @devrandom, @apoelstra and @TheBlueMatt about @sgeisler proposal before doing changes

devrandom commented:
It sounds plausible, but it's difficult to evaluate without trying it.  If it can be quickly prototyped, I can try it out in rust-bitcoin.

BTW, I had the same experience as @sgeisler when trying to apply the current version to rust-bitcoin, which motivated some of my previous questions...

TheBlueMatt commented:
Without having dug into it too much, building on @sgeisler's suggestion, what about:

```
trait ErrTrait { fn kind(&self) -> ErrorKind; }
impl Write { type Error: ErrTrait; ... }
impl ErrTrait for std::io::Error { .. }
```

RCasatta commented:
So I worked on the `ErrorKind` in 38bb903819033be2f9cc62f7b5b7cbcc3a58dfc9,  unfortunately, it is not in `core` so I had to duplicate the enum...

apoelstra commented:
I'm having trouble following all the discussion, especially as the PR is now 14 commits, many of which undo changes that earlier commits do. I agree with others saying that we should not use `dyn` anywhere. @sgeisler's suggested `IoError` still involves an allocation when constructing the error type which is far too much overhead for something that (conceptually) shouldn't need to be much more than a u16.

I'm also unconvinced of the merits of creating our own parallel implementation of the core2 crate. Our types which implement our traits will not be usable with other crates that expect the `core2` traits or vice-versa, and as we've seen there is actually a *significant* amount of complexity involved in simulating `Read`/`Wirte` well enough to make this usable. This is an extra maintenance burden that we don't want to take on, and it's very difficult to reason about all the potential feature-flag interactions.

TheBlueMatt commented:
> I'm also unconvinced of the merits of creating our own parallel implementation of the core2 crate

Ugh, yea, after all the work here, it just seems much more complicated than I was anticipating. Maybe if ErrorKind was exposed it could have gone down this path without the crazy amount of work here.
